src/internal/m365/exchange/events_restore.go

```diff
@@ -102,8 +108,115 @@ func (h eventRestoreHandler) restore(
 		return nil, clues.Stack(err)
 	}
 
+	// Fix up event instances in case we have a recurring event
+	err = h.updateRecurringEvents(ctx, userID, destinationID, ptr.Val(item.GetId()), event)
+	if err != nil {
+		return nil, clues.Stack(err)
+	}
+
 	info := api.EventInfo(event)
 	info.Size = int64(len(body))
 
 	return info, nil
 }
+
+func (h eventRestoreHandler) updateRecurringEvents(
+	ctx context.Context,
+	userID, containerID, eventID string,
+	event models.Eventable,
+) error {
+	if event.GetRecurrence() == nil {
+		return nil
+	}
+
+	// Cancellations and exceptions are currently in additional data
+	// but will get their own fields once the beta API lands and
+	// should be moved then
+	cancelledOccurrences := event.GetAdditionalData()["cancelledOccurrences"]
+	exceptionOccurrences := event.GetAdditionalData()["exceptionOccurrences"]
+
+	// OPTIMIZATION: Group instances whose dates are close by
+	if cancelledOccurrences != nil {
```

Comment:
I'd suggest moving the internals of cancelledOcurrences and exceptionOccurences to their own func, for ease of reading.  That'd shorten this func to something like below without much fuss, and allow you to break apart the inner sections for easier and more isolated unit testing.
```
{
  if getRecurrence == nil { return }
  
  cancelled, exception := additionalData(...)
  
  if cancelled != nil {
    handleCanceled()
  }
  
  if exception != nil {
    handleException()
  }
}
```